### PR TITLE
Helm acceptance tests

### LINKED ISSUE #1917: Helm acceptance tests
**Problem**
During Helm deployments we have no guarantee that the mirror node is healthy end to end.

**Solution**
Run the acceptance tests as a Helm test post deployment
- Add a `charts/hedera-mirror/templates/tests/pod.yaml` that runs the `hedera-mirror-test` image
- Add a `charts/hedera-mirror/templates/tests/secret.yaml` that stores the `application.yaml` config for the test pod
- Add test properties to `values.yaml` that allow to enable/disable, set image, config, etc. Disabled by default.

**Alternatives**

**Additional Context**

### PR DESCRIPTION
**Pull request by ijungmann**
**Detailed description**:
Add the ability to run acceptance tests to the Helm chart.

- Add a pod template to run the mirror node acceptance tests
- Add a secret template to hold the acceptance tests application.yml
- Update documentation to explain new helm tests.

**Which issue(s) this PR fixes**:
Fixes #1917 

**Special notes for your reviewer**:

**Checklist**
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

File: charts/hedera-mirror/templates/tests/pod.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    "helm.sh/hook": test
+spec:
+  containers:
+    - name: {{ include "hedera-mirror.fullname" . }}-test
+      image: "{{ .Values.test.image.repository }}:{{ .Values.test.image.tag | default .Chart.AppVersion }}"
+      imagePullPolicy: {{ .Values.test.image.pullPolicy }}
+      command:
+        - /bin/bash
+        - -ec
+        - ./mvnw clean integration-test -pl hedera-mirror-test -P="acceptance-tests" -Dcucumber.filter.tags=@Acceptance -e
```
Comment (ijungmann) [inline charts/hedera-mirror/templates/tests/pod.yaml:18]: Can add tag filter as a config option

File: charts/hedera-mirror/templates/tests/secret.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+apiVersion: v1
+kind: Secret
+metadata:
+  labels: {{ include "hedera-mirror.labels" . | nindent 4 }}
+  name: {{ printf "%s-acceptance-test" .Release.Name }}
+  namespace: {{ include "hedera-mirror.namespace" . }}
+stringData:
+  HEDERA_MIRROR_TEST_ACCEPTANCE_EMITBACKGROUNDMESSAGES: "{{ .Values.test.emitBackgroundMessages }}"
+  HEDERA_MIRROR_TEST_ACCEPTANCE_EXISTINGTOPICNUM: "{{ .Values.test.existingTopicNum }}"
+  HEDERA_MIRROR_TEST_ACCEPTANCE_MIRRORNODEADDRESS: {{ printf "%s-grpc:5600" .Release.Name | quote}}
```
Comment (ijungmann) [inline charts/hedera-mirror/templates/tests/secret.yaml:11]: Need to confirm this is the correct way of doing this, GRPC tests have not passed locally but could also be an unrelated issue.

Comment (ijungmann) [inline charts/hedera-mirror/templates/tests/secret.yaml:11]: Tried to move this into a config option in values.yaml, wasn't able to get the Release Name replacement working (done in sub-charts?)

File: charts/hedera-mirror/templates/tests/pod.yaml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  containers:
+    - name: {{ include "hedera-mirror.fullname" . }}-test
+      image: "{{ .Values.test.image.repository }}:{{ .Values.test.image.tag | default .Chart.AppVersion }}"
+      imagePullPolicy: {{ .Values.test.image.pullPolicy }}
+      envFrom:
+        - secretRef:
+            name: {{ include "hedera-mirror.fullname" . }}-acceptance-test
+      env:
+        - name: cucumberflags
+          value: "@acceptance"
```
Comment (ijungmann) [inline charts/hedera-mirror/templates/tests/pod.yaml:20]: Swap with actual yaml value

Comment (ijungmann) [inline charts/hedera-mirror/templates/tests/pod.yaml:20]: Was having issues earlier getting the tests to use this when it was in the Secret.  Need to test more, seems to work here though.

File: charts/hedera-mirror/templates/tests/pod.yaml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+spec:
+  containers:
+    - name: {{ include "hedera-mirror.fullname" . }}-test
+      image: "{{ .Values.test.image.repository }}:{{ .Values.test.image.tag | default .Chart.AppVersion }}"
+      imagePullPolicy: {{ .Values.test.image.pullPolicy }}
+      envFrom:
+        - secretRef:
+            name: {{ include "hedera-mirror.fullname" . }}-acceptance-test
+      env:
+        - name: cucumberflags
```
Comment (ijungmann) [inline charts/hedera-mirror/templates/tests/pod.yaml:19]: See if all caps works

File: charts/hedera-mirror/templates/tests/secret.yaml [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+metadata:
+  labels: {{ include "hedera-mirror.labels" . | nindent 4 }}
+  name: {{ include "hedera-mirror.fullname" . }}-acceptance-test
+  namespace: {{ include "hedera-mirror.namespace" . }}
+type: Opaque
+stringData:
+  application.yaml: |-
+    {{- toYaml .Values.test.config | nindent 4 }}
+  HEDERA_MIRROR_TEST_ACCEPTANCE_MIRRORNODEADDRESS: {{- tpl (toYaml .Values.test.endpoints.mirrorNodeAddress) $ | nindent 4 }}
+  HEDERA_MIRROR_TEST_ACCEPTANCE_REST_BASEURL: {{- tpl (toYaml .Values.test.endpoints.restBaseUrl) $ | nindent 4 }}
```
Comment (ijungmann) [inline charts/hedera-mirror/templates/tests/secret.yaml:13]: May be a way to make this cleaner, but I need the `tpl` to make use of the Release Name (or so I understand it, maybe a better way)

File: charts/hedera-mirror/values.yaml [hunk 6]
```diff
@@ -259,6 +259,23 @@ rest:
...
+  image:
+    pullPolicy: IfNotPresent
+    repository: gcr.io/mirrornode/hedera-mirror-test
+    tag: ""  # Defaults to the chart's app version
+  config:
+    emitBackgroundMessages: false
+    existingTopicNum:
+    network: testnet
+    operatorId:
+    operatorKey:
```
Comment (ijungmann) [inline charts/hedera-mirror/values.yaml:274]: Not sure what to do about some of these defaults.

File: charts/hedera-mirror/values.yaml [hunk 7]
```diff
@@ -259,6 +259,27 @@ rest:
...
+        test:
+          acceptance:
+            existingTopicNum:
+            # grpc endpoint
+            mirrorNodeAddress: "{{ .Release.Name }}-grpc:5600"
+            network:
+            # Do not use use 0.0.2 or 0.0.50 for operator to ensure crypto transfers are not waived
+            operatorId:
+            operatorKey:
+            restBaseUrl: "http://{{ .Release.Name }}-rest/api/v1"
```
Comment (Nana-EC) [inline charts/hedera-mirror/values.yaml:275]: This should be `rest.baseUrl`.
Are the rest checks working with this?

File: charts/hedera-mirror/values.yaml [hunk 8]
```diff
@@ -259,6 +259,27 @@ rest:
...
+      mirror:
+        test:
+          acceptance:
+            existingTopicNum:
+            # grpc endpoint
+            mirrorNodeAddress: "{{ .Release.Name }}-grpc:5600"
+            network:
+            # Do not use use 0.0.2 or 0.0.50 for operator to ensure crypto transfers are not waived
+            operatorId:
+            operatorKey:
```
Comment (Nana-EC) [inline charts/hedera-mirror/values.yaml:274]: May want to add `retrieveAddressBook: ` with default of `false`,  as some pre-prod envs address books aren't updated as regularly as they should be yet.

File: charts/hedera-mirror/templates/tests/pod.yaml [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    helm.sh/hook-delete-policy: hook-failed,hook-succeeded
+spec:
+  containers:
+    - name: acceptance-tests
+      image: "{{ .Values.test.image.repository }}:{{ .Values.test.image.tag | default .Chart.AppVersion }}"
+      imagePullPolicy: {{ .Values.test.image.pullPolicy }}
+      command:
+        - 'sh'
+        - '-c'
+        - |
```
Comment (Nana-EC) [inline charts/hedera-mirror/templates/tests/pod.yaml:19]: nit: you could add a `#` comment here to cover your copy intent prior to the run command

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Looking good, some nits and suggestions

File: charts/hedera-mirror/values.yaml (hunk 7 above)
Comment (ijungmann) [reply charts/hedera-mirror/values.yaml]: Good catch, I just got lucky that the default was testnet.  I tested with this corrected and it works, also tested with a bad endpoint and it failed so I know it's using it now.

File: charts/hedera-mirror/values.yaml (hunk 8 above)
Comment (ijungmann) [reply charts/hedera-mirror/values.yaml:274]: Sure, added

File: charts/hedera-mirror/templates/tests/pod.yaml (hunk 9 above)
Comment (ijungmann) [reply charts/hedera-mirror/templates/tests/pod.yaml:20]: Added a comment here, as well as in the secret to explain using `application-default`

File: charts/hedera-mirror/templates/tests/pod.yaml [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+  labels: {{- include "hedera-mirror.labels" . | nindent 4 }}
+  name: {{ include "hedera-mirror.fullname" . }}-acceptance-test
+  namespace: {{ include "hedera-mirror.namespace" . }}
+  annotations:
+    helm.sh/hook: test
+    helm.sh/hook-delete-policy: hook-failed,hook-succeeded
+spec:
+  containers:
+    - name: acceptance-tests
+      image: "{{ .Values.test.image.repository }}:{{ .Values.test.image.tag | default .Chart.AppVersion }}"
```
Comment (steven-sheehy) [inline charts/hedera-mirror/templates/tests/pod.yaml:14]: This should also use the `global.image.tag`. You'll need to add a `global.image: {}` to the values.yaml.
```suggestion
      image: "{{ .Values.test.image.repository }}:{{ .Values.global.image.tag | default .Values.test.image.tag | default .Chart.AppVersion }}"
```

File: charts/hedera-mirror/templates/tests/secret.yaml [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+apiVersion: v1
+kind: Secret
+metadata:
+  labels: {{ include "hedera-mirror.labels" . | nindent 4 }}
+  name: {{ include "hedera-mirror.fullname" . }}-acceptance-test
+  namespace: {{ include "hedera-mirror.namespace" . }}
+  annotations:
+    helm.sh/hook: test
+    helm.sh/hook-delete-policy: hook-failed,hook-succeeded
+
```
Comment (steven-sheehy) [inline charts/hedera-mirror/templates/tests/secret.yaml:11]: nit: extra line

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: charts/hedera-mirror/templates/tests/secret.yaml (hunk 11 above)
Comment (ijungmann) [reply charts/hedera-mirror/templates/tests/secret.yaml]: Removed.

File: charts/hedera-mirror/templates/tests/pod.yaml (hunk 10 above)
Comment (ijungmann) [reply charts/hedera-mirror/templates/tests/pod.yaml]: Sure, added that.

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Review by Nana-EC — APPROVED** [review]
LGTM

**Pull request merged**